skbio/diversity/_block.py

```diff
@@ -0,0 +1,280 @@
+        triangle, complain loudly.
+
+    Returns
+    -------
+    list of tuple
+        The ID pairs to compute distances between.
+    """
+    # if identical, gather the upper triangle
+    if len(rids) == len(cids) and (rids == cids).all():
+        return [(i, j) for idx, i in enumerate(rids) for j in rids[idx+1:]]
+
+    # otherwise, grab pairwise combinations disregarding the diagonal
+    else:
+        if set(rids).intersection(set(cids)):
+            raise ValueError("Attempting to compute a lower triangle")
+        return [(i, j) for i in rids for j in cids if i != j]
+
+
+def _block_kwargs(**kwargs):
+    """Construct arguments describing a block to compute
+
+    Returns
+    -------
+    dict
+        The parameters for the block of the distance matrix to compute.
+    """
+    for row_ids, col_ids in _generate_id_blocks(kwargs['ids'], kwargs['k']):
+        id_pairs = _pairs_to_compute(row_ids, col_ids)
+        if id_pairs:
+            kw = kwargs.copy()
+            kw['id_pairs'] = id_pairs
+            kw['row_ids'] = row_ids
+            kw['col_ids'] = col_ids
+
+            yield kw
+
+
+def _block_compute(**kwargs):
+    """Compute a block within the resulting distance matrix
+
+    Notes
+    -----
+    This method encapsulates the two expensive operations to perform for each
+    block, namely, the "shearing" of the phylogenetic tree to correspond to
+    only the OTUs of interest, and the actual beta diversity calculations.
+
+    Returns
+    -------
+    DistanceMatrix
+    """
+    block_kw = _block_party(**kwargs)
+
+    return partial_beta_diversity(**block_kw)
+
+
+def _map(func, kw_gen):
+    """Map a function over arguments
+
+    Note
+    ----
+    builtin map does not allow for mapping with kwargs.
+
+    """
+    for kwargs in kw_gen:
+        yield func(**kwargs)
+
+
+def _reduce(blocks):
+    """Reduce an iterable of partial distance matrices into a full matrix"""
+    all_blocks = list(blocks)
+
+    # Determine the maximum integer ID observed in the blocks. There exists a
+    # 1-1 mapping between the integer ID and a sample ID. We increment by 1
+    # as the integer ID space begins with zero, and we'll be using this value
+    # to determine the size of the resulting full distance matrix.
+    n_ids = max(map(lambda x: max(x.ids), all_blocks)) + 1
+
+    mat = np.zeros((n_ids, n_ids), dtype=float)
+
+    # TODO: something smarter.
+    for block in all_blocks:
+        n_blk_ids = len(block.ids)
+
+        # get the corresponding coordinates in the master matrix
+        master_idx = [(i, j) for row, i in enumerate(block.ids)
+                      for j in block.ids[row+1:]]
+
+        # get the corresponding coordinates within the current block
+        block_idx = [(i, j) for row, i in enumerate(range(n_blk_ids))
+                     for j in range(row+1, n_blk_ids)]
+
+        for (m_i, m_j), (b_i, b_j) in zip(master_idx, block_idx):
+            mat[m_i, m_j] += block.data[b_i, b_j]
```

Comment:
Could this be vectorized so that you could just add `block.data` in one go? Since 0 represents an uncalculated element, I don't see why that wouldn't work.